src/main/java/org/prebid/server/analytics/pubstack/PubstackEventHandler.java

```diff
@@ -0,0 +1,198 @@
+        this.timeoutMs = pubstackAnalyticsProperties.getTimeoutMs();
+        this.jacksonMapper = Objects.requireNonNull(jacksonMapper);
+        this.httpClient = Objects.requireNonNull(httpClient);
+        this.vertx = Objects.requireNonNull(vertx);
+
+        this.lockOnSend = new ReentrantLock();
+        this.events = new AtomicReference<>(new ConcurrentLinkedQueue<>());
+        this.headers = makeHeaders();
+        this.byteSize = new AtomicLong();
+        if (enabled) {
+            this.reportTimerId = setReportTtlTimer();
+        }
+    }
+
+    public <T> void handle(T event) {
+        if (enabled) {
+            buffer(event);
+            reportEventsOnCondition(byteSize -> byteSize.get() > maxByteSize, byteSize);
+            reportEventsOnCondition(eventsReference -> eventsReference.get().size() > maxEventCount, events);
+        }
+    }
+
+    public void reportEvents() {
+        if (enabled) {
+            reportEventsOnCondition(events -> events.get().size() > 0, events);
+        }
+    }
+
+    public void updateConfig(boolean enabled, String endpoint, String scopeId) {
+        updateTimerOnEnabling(enabled);
+        this.enabled = enabled;
+        this.endpoint = endpoint;
+        this.scopeId = scopeId;
+    }
+
+    private <T> void buffer(T event) {
+        final ObjectNode eventNode = jacksonMapper.mapper().valueToTree(event);
+        eventNode.put(SCOPE_FIELD_NAME, scopeId);
+        final String jsonEvent = jacksonMapper.encode(eventNode);
+        events.get().add(jsonEvent);
+        byteSize.getAndAdd(jsonEvent.getBytes().length);
+    }
+
+    private <T> boolean reportEventsOnCondition(Predicate<T> conditionToSend, T conditionValue) {
+        boolean requestWasSent = false;
+        if (conditionToSend.test(conditionValue)) {
+            lockOnSend.lock();
+            try {
+                if (conditionToSend.test(conditionValue)) {
+                    requestWasSent = true;
+                    sendEvents(events);
+                }
+            } catch (Exception exception) {
+                logger.error("[pubstack] Failed to send analytics report to endpoint {0} with a reason {1}",
+                        endpoint, exception.getMessage());
+            } finally {
+                lockOnSend.unlock();
+            }
+        }
+        return requestWasSent;
+    }
+
+    private void sendEvents(AtomicReference<Queue<String>> events) {
+        final Queue<String> copyToSend = events.get();
+        resetReportEventsConditions(events);
+        httpClient.request(HttpMethod.POST, HttpUtil.validateUrl(endpoint), headers, toGzippedBytes(copyToSend),
+                timeoutMs)
+                .setHandler(this::handleReportResponse);
+    }
+
+    private void resetReportEventsConditions(AtomicReference<Queue<String>> events) {
+        events.set(new ConcurrentLinkedQueue<>());
+        byteSize.set(0);
+        vertx.cancelTimer(reportTimerId);
+        reportTimerId = setReportTtlTimer();
```

Comment:
Not sure if these are thread safe. The rest of the code seems to be implemented in a thread safe way, so this should probably be too.

Maybe calculating the `byteSize` with the copied queue is a better way than having two states that can separately be set.

Reply:
Since `ReentrantLock` is used while sending the events, imho it is quite safe to use those in such way.